api/libproject/generate_appengine_cron_yaml.rb

```diff
@@ -15,9 +17,32 @@ def to_cron_path(suffix)
   return "#{CRON_DIR}/cron_#{suffix}.yaml"
 end
 
-from_path = to_cron_path("default")
-if File.file?(to_cron_path(aou_env))
-  from_path = to_cron_path(aou_env)
+def merge_cron_yaml(base_yaml_file, env_yaml_file, output_yaml_file)
+  common = Common.new
+
+  out_yaml = YAML.load_file(base_yaml_file)
+  base_crons = out_yaml['cron'].map { |c| [c['url'], c] }.to_h
+
+  if File.file?(env_yaml_file)
+    env_yaml = YAML.load_file(env_yaml_file)
+    env_yaml['cron'].each do |env_cron|
+
+      raise ArgumentError.new("missing required 'url' key for #{env_cron}") unless env_cron.key?('url')
+      url = env_cron['url']
+      raise ArgumentError.new("cron URL '#{url}' does not exist in base yaml") unless base_crons.key?(url)
+
+      if env_cron.key?('aou_skip_reason')
+        common.status "Excluding cron '#{url}', reason: '#{env_cron['aou_skip_reason']}'"
+        out_yaml['cron'].reject! {|o| o['url'] == url}
+        next
+      end
+
+      # This indirectly mutates out_yaml
+      base_crons[url].merge!(env_cron)
```

Comment:
does mutating the out_yaml directly not work? `out_yaml['cron'][url]`

I think it's a little easier to reason about if the code block above this and this line mutate the same object when it wants to make changes.

Alternatively, we could just remove this line? It's not actually being exercised right now right? The current configs only utilize the removal part of this cron merger.

Reply:
also, I have a slight preference for rewriting these with the non `!` functions. 

Reason being, I think less people on our team would know what that does by just looking at it whereas they would with the immutable version `base_crons[url] = base_crons[url].merge(env_cron)`. Also, I think we generally lean towards the immutability pattern when possible in the UI/API codebase.

Even if the mutating pattern is better practice in Ruby, there's some benefit in having it consistent with the patterns we see elsewhere.

Reply:
RE mutating out_yaml:
Per package comment, I'm treating the cron URL as the primary key. However, it's not actually a key in a map, which was why I created an additional data structure here i.e. `out_yaml['cron']` is an array, not a dict.

RE removing functionality:
The previous codebase did utilize a schedule override in perf. As soon as the freeTierBilling cron is fixed, we're likely to use that again. I'd rather have the complete tool in place so that I'm not removing functionality from the previous version (previous version was very verbose, but it did allow you to overwrite anything) - the next person who comes along is not likely to have the appetite to modify this tool.

RE mutability: yeah, this was just laziness and lack of confidence in the yaml package documentation. Rewrote to avoid this pattern